packages/core/content-manager/server/src/services/document-metadata.ts

```diff
@@ -56,45 +69,56 @@ export default ({ strapi }: { strapi: Strapi }) => ({
    */
-  async getAvailableLocales(uid: Common.UID.ContentType, document: DocumentVersionSelector) {
-    if (!document.locale) return [];
-
-    const status = document.publishedAt !== null ? 'published' : 'draft';
-
-    // TODO: Use document service instead of query engine
-    // Find other locales of the document in the same status
-    return strapi.db.query(uid).findMany({
-      where: {
-        documentId: document.id,
-        // Omit current one
-        locale: { $ne: document.locale },
-        // Find locales of the same status
-        publishedAt: status === 'published' ? { $ne: null } : null,
-      },
-      select: ['id', 'locale', 'updatedAt', 'createdAt', 'publishedAt'],
-    }) as unknown as DocumentMetadata['availableLocales'];
-  },
+  async getAvailableLocales(version: DocumentVersion, allVersions: DocumentVersion[]) {
+    // Group all versions by locale
+    const versionsByLocale = groupBy('locale', allVersions);
 
-  async getAvailableStatus(uid: Common.UID.ContentType, document: DocumentVersionSelector) {
-    // Find if the other status of the document is available
-    const status = document.publishedAt !== null ? 'published' : 'draft';
-    const otherStatus = status === 'published' ? 'draft' : 'published';
+    // Delete the current locale
+    delete versionsByLocale[version.locale];
 
-    return strapi.documents(uid).findOne(document.id, {
-      // TODO: Do not filter by locale if i18n is disabled
-      locale: document.locale,
-      status: otherStatus,
-      fields: ['id', 'updatedAt', 'createdAt', 'publishedAt'],
-    }) as unknown as DocumentMetadata['availableStatus'][0] | null;
+    // For each locale, get the ones with the same status
+    return Object.values(versionsByLocale).map((localeVersions: DocumentVersion[]) => {
+      const draftVersion = localeVersions.find((v) => v.publishedAt === null);
+      const otherVersions = localeVersions.filter((v) => v.id !== draftVersion?.id);
+
+      if (!draftVersion) return;
+
+      return {
+        ...pick(AVAILABLE_LOCALES_FIELDS, draftVersion),
+        status: this.getStatus(draftVersion, otherVersions as any),
+      };
+    });
   },
 
+  /**
+   * Returns available status of a document for the current locale
+   */
+  async getAvailableStatus(version: DocumentVersion, allVersions: DocumentVersion[]) {
+    // Find the other status of the document
+    const status =
+      version.publishedAt !== null
+        ? CONTENT_MANAGER_STATUS.DRAFT
+        : CONTENT_MANAGER_STATUS.PUBLISHED;
+
+    // Get version that match the current locale and not match the current status
+    const availableStatus = allVersions.find((v) => {
+      const matchLocale = v.locale === version.locale;
+      const matchStatus = status === 'published' ? v.publishedAt !== null : v.publishedAt === null;
+      return matchLocale && matchStatus;
+    });
+
+    if (!availableStatus) return availableStatus;
+
```

Comment:
```suggestion
    return { ...pick(AVAILABLE_STATUS_FIELDS, availableStatus), status };
```

Reply:
indeed, will add this in another PR, as I wanted to discuss something with josh about this!